File: pa2/Lex.c
//-----------------------------------------------------------------------------
// Carlos Barillas - cbarilla@ucsc.edu
// SID: 1380581
// CMPS101
// PA2 - Lex.c
//-----------------------------------------------------------------------------
#include<stdio.h>
#include<stdlib.h>
#include<string.h>

#include "List.h"

#define MAX_LEN 256

int main(int argc, char* argv[]) {
  if (argc != 3) {
    printf("Usage: %s <input file> <output file>\n", argv[0]);
    exit(1);
  }

  int numLines = 0;
  char line[MAX_LEN];

  FILE* in = fopen(argv[1], "r");
  FILE* out = fopen(argv[2], "w");

  // counts number of lines in input files
  while (fgets(line, MAX_LEN, in ) != NULL) {
    numLines++;
  }

  // reset file pointer to start of file
  rewind(in);

  // places the strings in file into an array
  int n = 0;
  char* array[numLines];
  for(int index = 0; index<=numLines;index++){
    array[index] = calloc(1, sizeof(char* [MAX_LEN]) );
   }
  while(fgets(line, MAX_LEN, in) != NULL){
    strcpy(array[n], line );
    n++;
   }

  // insertion sort algorithm
  List myList = newList();
  
  append(myList, 0);
  for (int i = 0; i < numLines - 1; i++) {
    moveFront(myList);
    for (int j = -i; j < 1; j++) {
      if (strcmp(array[get(myList)], array[i + 1]) > 0) {
        insertBefore(myList, i + 1);
        j = 1;
      } else if (index(myList) == length(myList) - 1) {
        append(myList, i + 1);
        j = 1;
      } else
        moveNext(myList);
    }
  }

  // print list to out file
  moveFront(myList);
  while (index(myList) != -1) {
    fprintf(out, "%s", array[get(myList)]);
    moveNext(myList);
  }

  fclose(in);
  fclose(out);

  // free memory
  int m = 0;
  while( m <= numLines) {
    free(array[m]);
    m++;
  }
  freeList(&myList);
}

File: pa1/Lex.java
/*
 * Carlos Barillas - cbarilla@ucsc.edu
 * CMPS101 - PA1 - Lex.java
 *  
 * Takes two command line arguments. Input file can be any text file, output 
 * file will be the input arranged in alphabetical order. Uses Insertion Sort
 * to indirectly sort the array by building a List of indices in a certain order.
 *
 */


import java.io.*;
import java.util.Scanner;

public class Lex {
  public static void main(String[] args) throws IOException {
      
    // Declaring variables
    Scanner in = null;
    PrintWriter out = null;
    String line = null;
    String[] array = null;
    List list = new List();
    int lines = 0;
      
    // Check if two arguments are passed
    if(args.length != 2) {
      System.err.println("Usage: Lex infile outfile");
      System.exit(1);
    }

    // Gets file from command line arg
    in = new Scanner(new File(args[0]));

    // Counts lines of file
    while(in.hasNextLine()) {
      ++lines;
      in.nextLine();
    }

    in.close();
    in = null;

    // Allocate enough space in an array
    array = new String[lines];
    // Scanning both the input file and output file
    in = new Scanner(new File(args[0]));
    out = new PrintWriter(new FileWriter(args[1]));

    // Add lines from file into array
    for (int i = 0; i < lines; i++){
      line = in.nextLine()+" ";   
      array[i] = line;
    } 		        		

    // Put first line of the input file into list (already sorted)
    list.append(0);
    // Insertion Sort Algorithm
    for (int i = 0; i < lines-1; i++) {
      list.moveFront();
      for (int j = -i ; j < 1; j++) {
        if (array[list.get()].compareTo(array[i+1])>0) {  
    	  list.insertBefore(i+1);
          j = 1;
        }
        else if (list.index() == list.length()-1) {
    	  list.append(i+1);
          j = 1;
        }		 
        else 
        list.moveNext(); 		  
      }  
    }

    // Reset index to the front of the List
    // Prints sorted list.
    list.moveFront();
    while(list.index() >= 0) {
      out.println(array[list.get()]);
      list.moveNext();
    }
    in.close();
    out.close();
  }
}
